### PR TITLE
JIT: Improve call args interference checking when stores are involved

### LINKED ISSUE #13758: Missing `GTF_ASG` and `GTF_ALL_EFFECT` on GenTree nodes.
Sometimes we are missing `GTF_ASG` flag for calls when create a setup argument node in `fgArgInfo::EvalArgsToTemps()`. It creates an unwanted effect that call to `gtUpdateSideEffects(tree, stmt)` could change flags when the trees are unchanged:

before calling `gtUpdateSideEffects(000055, stmt)`:
```

N019 ( 51, 40) [000068] --CXG-------              *  JTRUE     void
N018 ( 49, 38) [000067] J-CXG--N----              \--*  LT        int    $148
N016 ( 47, 36) [000065] --CXG------- HERE            +--*  CALLV stub int    System.Collections.IComparer.Compare $383
N010 ( 18, 11) [000131] -ACXG---R-L- arg2 SETUP      |  +--*  ASG       ref    $346
N009 (  1,  1) [000130] D------N----                 |  |  +--*  LCL_VAR   ref    V12 tmp5         d:2 $346
N008 ( 18, 11) [000059] --CXG-------                 |  |  \--*  CALL r2r_ind ref    System.Array.GetValue $346
N006 (  3,  2) [000049] *--XG------- this in rcx     |  |     +--*  IND       ref    <l:$286, c:$285>
N005 (  1,  1) [000048] ------------                 |  |     |  \--*  LCL_VAR   byref  V00 this         u:1 Zero Fseq[keys] $80
N007 (  1,  1) [000055] ------------ arg1 in rdx     |  |     \--*  LCL_VAR   int    V05 loc2         u:5 (last use) $147
N012 (  1,  1) [000132] ------------ arg2 in rdx     |  +--*  LCL_VAR   ref    V12 tmp5         u:2 (last use) $346
N013 (  1,  1) [000064] ------------ arg3 in r8      |  +--*  LCL_VAR   ref    V04 loc1         u:2 $205
N014 (  1,  1) [000062] ------------ this in rcx     |  +--*  LCL_VAR   ref    V09 tmp2         u:2 (last use) <l:$241, c:$341>
N015 (  3, 10) [000127] ------------ arg1 in r11     |  \--*  CNS_INT(h) long   0xd1ffab1e ftn REG r11 $480
N017 (  1,  1) [000066] ------------                 \--*  CNS_INT   int    0 $40
```
after:

```
N019 ( 51, 40) [000068] -ACXG-------              *  JTRUE     void
N018 ( 49, 38) [000067] JACXG--N----              \--*  LT        int    $148
N016 ( 47, 36) [000065] -ACXG------- HERE                        +--*  CALLV stub int    System.Collections.IComparer.Compare $383
N010 ( 18, 11) [000131] -ACXG---R-L- arg2 SETUP      |  +--*  ASG       ref    $346
N009 (  1,  1) [000130] D------N----                 |  |  +--*  LCL_VAR   ref    V12 tmp5         d:2 $346
N008 ( 18, 11) [000059] --CXG-------                 |  |  \--*  CALL r2r_ind ref    System.Array.GetValue $346
N006 (  3,  2) [000049] *--XG------- this in rcx     |  |     +--*  IND       ref    <l:$286, c:$285>
N005 (  1,  1) [000048] ------------                 |  |     |  \--*  LCL_VAR   byref  V00 this         u:1 Zero Fseq[keys] $80
N007 (  1,  1) [000055] ------------ arg1 in rdx     |  |     \--*  LCL_VAR   int    V05 loc2         u:5 (last use) $147
N012 (  1,  1) [000132] ------------ arg2 in rdx     |  +--*  LCL_VAR   ref    V12 tmp5         u:2 (last use) $346
N013 (  1,  1) [000064] ------------ arg3 in r8      |  +--*  LCL_VAR   ref    V04 loc1         u:2 $205
N014 (  1,  1) [000062] ------------ this in rcx     |  +--*  LCL_VAR   ref    V09 tmp2         u:2 (last use) <l:$241, c:$341>
N015 (  3, 10) [000127] ------------ arg1 in r11     |  \--*  CNS_INT(h) long   0xd1ffab1e ftn REG r11 $480
N017 (  1,  1) [000066] ------------                 \--*  CNS_INT   int    0 $40
```
see that flags on [000065] have changed, but they should not.

The same happens with `GTF_ALL_EFFECT` on addr modes.

The fix for it is simple and published in dotnet/coreclr#27732, however, the fix create 0.1% size regression because of another issue. So I will postpone it until another issue is fixed.

Along the way fix that we don't print `fgMorphTree block, stmt(after)` if only the flags were changed.

category:correctness
theme:ir
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by jakobbotsch**
- Fix the propagation of `GTF_ASG` during call args morphing
- Introduce `Compiler::gtMayHaveStoreInterference` that can check whether two trees interfere with each other due to a store in one tree that stores to a local read by the other tree
- Use the new helper when checking for whether we should reverse `GT_STOREIND` nodes
- Use the new helper when deciding whether previous args need to be evaluated to temps because we see an argument with an embedded store (typically a call now that we propagate flags correctly).
- Use the new helpers when checking interference in `optRedundantRelop`

Fix #13758

### CODE REVIEW

**Comment by jakobbotsch** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=537723&view=ms.vss-build-web.run-extensions-tab) are kind of large, especially MinOpts ones. That's because with the correct flags propagation we are now much more conservative about evaluating call arguments into temps if those arguments are themselves calls that ended up with embedded stores. I'm going to try enabling the full fledged interference check in MinOpts to see what the impact is (it has large size reductions, so perhaps the TP cost is balanced out by the IR reduction).

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=538729&view=ms.vss-build-web.run-extensions-tab). Wins across the board from enabling this optimization and also doing it in MinOpts.

Initially I just wanted to fix #13758. However, fixing the propagation causes several regressions in places that are using `GTF_ASG` for interference checks. I identified these three places
- Call args morphing
- `optRedundantRelop`
- `gtSetEvalOrder` in the `GT_STOREIND` case

and switched them to more precise checking. The first seems to be by far the most impactful.

I've looked at most of the benchmarks.run_pgo FullOpts regressions. Almost all of them are because the removal of these temps means less live range splitting for LSRA. This doesn't usually result in new spills, but it does mean we sometimes pick registers that take more space to encode, or where we end up needing another reg-reg mov. For example, most regressions are similar in spirit to the following:

```diff
 ; Assembly listing for method Microsoft.Cci.MetadataWriter:GetTypeAttributes(Microsoft.Cci.ITypeDefinition):int:this (Tier1)
 ; Emitting BLENDED_CODE for X64 with AVX512 - Windows
 ; Tier1 code
 ; optimized code
 ; rsp based frame
 ; partially interruptible
 ; No matching PGO data
 ; Final local variable assignments
 ;
 ;  V00 this         [V00,T00] (  3,  3   )     ref  ->  rcx         this class-hnd single-def <Microsoft.Cci.MetadataWriter>
-;  V01 arg1         [V01,T01] (  3,  3   )     ref  ->  rdx         class-hnd single-def <Microsoft.Cci.ITypeDefinition>
+;  V01 arg1         [V01,T01] (  3,  3   )     ref  ->  rax         class-hnd single-def <Microsoft.Cci.ITypeDefinition>
 ;  V02 OutArgs      [V02    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
 ;  V03 tmp1         [V03    ] (  2,  4   )  struct (48) [rsp+0x28]  do-not-enreg[XS] addr-exposed "by-value struct argument" <Microsoft.CodeAnalysis.Emit.EmitContext>
-;  V04 tmp2         [V04,T02] (  2,  4   )     ref  ->  rdx         single-def "argument with side effect"
 ;
 ; Lcl frame size = 88
 
 G_M23091_IG01:        ; bbWeight=1, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, nogc <-- Prolog IG
        sub      rsp, 88
        vzeroupper 
-						;; size=7 bbWeight=1 PerfScore 1.25
-G_M23091_IG02:        ; bbWeight=1, gcrefRegs=0006 {rcx rdx}, byrefRegs=0000 {}, byref, nogc
-       ; gcrRegs +[rcx rdx]
+       mov      rax, rdx
+       ; gcrRegs +[rax]
+						;; size=10 bbWeight=1 PerfScore 1.50
+G_M23091_IG02:        ; bbWeight=1, gcrefRegs=0003 {rax rcx}, byrefRegs=0000 {}, byref, nogc
+       ; gcrRegs +[rcx]
        vmovdqu  ymm0, ymmword ptr [rcx+0xD0]
        vmovdqu  ymmword ptr [rsp+0x28], ymm0
        vmovdqu  xmm0, xmmword ptr [rcx+0xF0]
        vmovdqu  xmmword ptr [rsp+0x48], xmm0
 						;; size=28 bbWeight=1 PerfScore 11.00
 G_M23091_IG03:        ; bbWeight=1, extend
-       mov      rcx, rdx
        lea      rdx, [rsp+0x28]
-       ; gcrRegs -[rdx]
+       mov      rcx, rax
        call     [<unknown method>]
-       ; gcrRegs -[rcx]
+       ; gcrRegs -[rax rcx]
        ; gcr arg pop 0
        nop      
 						;; size=15 bbWeight=1 PerfScore 4.00
 G_M23091_IG04:        ; bbWeight=1, epilog, nogc, extend
        add      rsp, 88
        ret      
 						;; size=5 bbWeight=1 PerfScore 1.25
 
-; Total bytes of code 55, prolog size 7, PerfScore 17.50, instruction count 12, allocated bytes for code 55 (MethodHash=af30a5cc) for method Microsoft.Cci.MetadataWriter:GetTypeAttributes(Microsoft.Cci.ITypeDefinition):int:this (Tier1)
+; Total bytes of code 58, prolog size 7, PerfScore 17.75, instruction count 13, allocated bytes for code 58 (MethodHash=af30a5cc) for method Microsoft.Cci.MetadataWriter:GetTypeAttributes(Microsoft.Cci.ITypeDefinition):int:this (Tier1)
 ; ============================================================
```
which results in the one additional reg-reg mov.

File: src/coreclr/jit/redundantbranchopts.cpp [hunk 1]
```diff
@@ -2148,7 +2139,7 @@ bool Compiler::optRedundantRelop(BasicBlock* const block)
 
         for (unsigned int i = 0; i < definedLocalsCount; i++)
         {
-            if (gtHasRef(prevTreeData, definedLocals[i]))
+            if (gtTreeHasLocalRead(prevTreeData, definedLocals[i]))
```
Comment (jakobbotsch) [inline src/coreclr/jit/redundantbranchopts.cpp:2142]: Also a small bug fix here -- `gtHasRef` does not take promotion into account.

**Review by AndyAyersMS — COMMENTED** [review]
Changes look good.

I wonder if it would be more efficient to generalize this to sets of locals? You could have the same kind of cap where if the sets grow too large you just bail out.

**Review by AndyAyersMS — APPROVED** [review]
Meant to approve, comments above are for possible follow up.

**Comment by jakobbotsch** [comment]
> Changes look good.
> 
> I wonder if it would be more efficient to generalize this to sets of locals? You could have the same kind of cap where if the sets grow too large you just bail out.

Seems likely it would be, e.g. I could factor `SmallValueNumSet` to a general set type and try to reuse it here. I'll consider that for a follow up.
We also have `ALLVARSET`, but I dislike it given how it pushes the handling of cases where you have more than `lclMAX_ALLSET_TRACKED` locals onto the user.

**Comment by cincuranet** [comment]
Improvements:
- https://github.com/dotnet/perf-autofiling-issues/issues/28379

**Pull request merged**